ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +579,113 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
     );
   }
 
+  private void verifyJoin(final Join root, final PlanNode left, final PlanNode right) {
+    if (left.getNodeOutputType() == DataSourceType.KSTREAM) {
+      if (right.getNodeOutputType() == DataSourceType.KTABLE) {
+
+        // stream-table join detected
+
+        if (root.getInfo().getType().equals(JoinType.OUTER)) {
+          throw new KsqlException("Invalid join type:"
+              + " full-outer join not supported for stream-table joins.");
+        }
+
+        if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+          throw new KsqlException("Invalid join condition:"
+              + " stream-table joins require to join on the table's primary key.");
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (right.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException("Invalid join order:"
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the right input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(root.getInfo().getLeftJoinExpression(), left)) {
+        // after we lift the n-way join restriction, we should be able to support FK-joins
+        // at any level in the join tree, even after we add right-deep/bushy join tree support,
+        // because a FK-join output table has the same PK as its left input table
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final Expression joinExpression,
```

Comment:
without looking into the code, it feels to me like this should be part of `JoinTree` logic (and we should output this information in `JoinTree.Node`). Ideally, anything that has to do with "resolving" a Join should be outside of LogicalPlanner since this class is already getting a little unwieldly before this change - and also I think we can leverage the recursion used there to make this a little cleaner

Reply:
+10 here.

Reply:
I agree that we might want to do some refactoring, but it should not be part of this PR IMHO.

In particular, the logical planner should use a visitor pattern IMHO to build the logical plan. Just pushing out "random methods" into "random helper" classed, does make `LogicalPlanner` shorter, but does not really improve the code structure IMHO.

Reply:
to defend my original comment 😝 I don't think `JoinTree` is a random helper, it already has the logic to extract keys and traverse N-way join trees. That's why I thought this would fit in there. Up to you if you feel there's a better way and when you want to do it (which PR)

Reply:
I keep it in mind but don't want to hold off this PR any longer, as it blocks follow up PRs for FK-joins.